### PR TITLE
[MM-31132] app/server: add pprof endpoint

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Move profiling and metrics server initialization to server.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-31132

#### Release Note
```release-note
Profiling mattermost server with pprof is now available for Team Edition.
```

### CODE REVIEW

File: app/config.go [hunk 1]
```diff
@@ -411,12 +411,10 @@ func (s *Server) SaveConfig(newCfg *model.Config, sendConfigChangeClusterMessage
 		return model.NewAppError("saveConfig", "app.save_config.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	if s.Metrics != nil {
-		if *s.Config().MetricsSettings.Enable {
-			s.Metrics.StartServer()
-		} else {
-			s.Metrics.StopServer()
-		}
+	if *s.Config().MetricsSettings.Enable {
```
Comment (jespino) [inline app/config.go:414]: should we check the `s.startMetrics` attribute here too?

Comment (jespino) [reply app/config.go:414]: Should we call the `s.Metrics.Register` function too? what happens if I have disabled the metrics at start time but I enable it later?

Comment (agnivade) [reply app/config.go:414]: Same thoughts.

File: app/server.go [hunk 2]
```diff
@@ -1454,6 +1465,93 @@ func doCheckWarnMetricStatus(a *App) {
...
+	s.metricsRouter = mux.NewRouter()
+	runtime.SetBlockProfileRate(*s.Config().MetricsSettings.BlockProfileRate)
+
+	rootHandler := func(w http.ResponseWriter, r *http.Request) {
+		html := `
+			<html>
+				<body>
+					<div><a href="/metrics">Metrics</a></div>
+					<div><a href="/debug/pprof/">Profiling Root</a></div>
+					<div><a href="/debug/pprof/cmdline">Profiling Command Line</a></div>
```
Comment (agnivade) [inline app/server.go:1499]: While we are here, let's add the /profile and /trace endpoints in the HTML response.

File: app/server.go [hunk 3]
```diff
@@ -1454,6 +1465,93 @@ func doCheckWarnMetricStatus(a *App) {
...
+
+	if s.metricsServer != nil {
+		s.metricsServer.Close()
+		s.Log.Info("Metrics and profiling server is stopping", mlog.String("address", *s.Config().MetricsSettings.ListenAddress))
+		s.metricsServer = nil
+	}
+}
+
+func (s *Server) HandleMetrics(route string, h http.Handler) {
+	if s.metricsRouter != nil {
```
Comment (agnivade) [inline app/server.go:1480]: Is there a reason for this check? `HandleMetrics` is anyways called after `InitMetricsRouter`.

File: app/server.go [hunk 4]
```diff
@@ -1454,6 +1465,93 @@ func doCheckWarnMetricStatus(a *App) {
...
+		s.metricsRouter.Handle(route, h)
+	}
+}
+
+func (s *Server) InitMetricsRouter() {
+	if s.metricsRouter != nil {
+		return
+	}
```
Comment (agnivade) [inline app/server.go:1488]: Are there cases where this method is called multiple times?

File: app/server.go [hunk 5]
```diff
@@ -1454,6 +1465,93 @@ func doCheckWarnMetricStatus(a *App) {
 	}
 }
 
+func (s *Server) StopMetricsServer() {
+	s.metricsLock.Lock()
+	defer s.metricsLock.Unlock()
+
+	if s.metricsServer != nil {
+		s.metricsServer.Close()
```
Comment (agnivade) [inline app/server.go:1473]: Let's use the `Shutdown` method instead of `Close`.

**Review by agnivade — COMMENTED** [review]
Just some minor comments.

File: app/server.go (hunk 3 above)
Comment (isacikgoz) [reply app/server.go:1480]: It was for a safety measure since this method is called from ee repo as well. And sometimes it's hard to follow code from there. 0/5 to leave it.

File: app/config.go (hunk 1 above)
Comment (isacikgoz) [reply app/config.go:414]: Yes we should use `s.startMetrics` and also +1 to registering handlers here.

File: app/server.go (hunk 4 above)
Comment (isacikgoz) [reply app/server.go:1488]: Actually we can remove it 👍

File: app/config.go [hunk 6]
```diff
@@ -411,12 +411,13 @@ func (s *Server) SaveConfig(newCfg *model.Config, sendConfigChangeClusterMessage
...
 	}
 
-	if s.Metrics != nil {
-		if *s.Config().MetricsSettings.Enable {
-			s.Metrics.StartServer()
-		} else {
-			s.Metrics.StopServer()
+	if s.startMetrics && *s.Config().MetricsSettings.Enable {
+		if s.Metrics != nil {
+			s.Metrics.Register()
```
Comment (jespino) [inline app/config.go:416]: Is register idempotent? or what is the same, is the router `Handle` method idempotent?

File: app/server.go [hunk 7]
```diff
@@ -1454,6 +1463,90 @@ func doCheckWarnMetricStatus(a *App) {
...
+
+func (s *Server) InitMetricsRouter() {
+	s.metricsRouter = mux.NewRouter()
+	runtime.SetBlockProfileRate(*s.Config().MetricsSettings.BlockProfileRate)
+
+	rootHandler := func(w http.ResponseWriter, r *http.Request) {
+		html := `
+			<html>
+				<body>
+					<div><a href="/metrics">Metrics</a></div>
```
Comment (jespino) [inline app/server.go:1492]: nit: should this be conditionally shown?

**Review by jespino — COMMENTED** [review]
A couple of the details, but otherwise looks good to me.

File: app/config.go (hunk 6 above)
Comment (isacikgoz) [reply app/config.go:416]: Yes, it is. Register will use same `prometheus.Registry`.  As far as I see the `Handle` method will be idempotent in this case as we don't modify the route thereafter.

**Review by jespino — APPROVED** [review]
LGTM

File: app/server.go (hunk 3 above)
Comment (agnivade) [reply app/server.go]: Sounds good. Let's keep it then.

**Review by agnivade — APPROVED** [review]
Neat work!

File: app/server.go (hunk 7 above)
Comment (saturninoabril) [reply app/server.go]: +1 to conditionally show if accessible. In TE, it's shown then throw 404 on click/visit.

**Review by saturninoabril — COMMENTED** [review]
Tested and works as expected, but observed few things:
1. `/admin_console/environment/performance_monitoring` is not accessible on TE. Should it be made accessible in sysadmin page for TE? If yes, is there a corresponding PR to webapp?
2. It requires server restart to take effect access or restrict to `/metrics`  when removing or adding E20 license.

Let me know how do you want to proceed.

**Comment by isacikgoz** [comment]
Thanks for the thorough review @saturninoabril 💪 I added the webapp PR, I completely forgot it. As for the 2. item you've pointed out it's a very good point. I'll update the EE PR to add/remove handler on license change.

**Review by saturninoabril — APPROVED** [review]
Thanks @isacikgoz for updating! Tested and works as expected.
